batcher/sweep_batcher.go

```diff
@@ -0,0 +1,591 @@
+	// sweep funds.
+	musig2ServerSign MuSig2SignSweep
+
+	// verifySchnorrSig is a function that can be used to verify a schnorr
+	// signature.
+	VerifySchnorrSig VerifySchnorrSig
+
+	// chainParams are the chain parameters of the chain that is used by
+	// batches.
+	chainParams *chaincfg.Params
+
+	// store includes all the database interactions that are needed by the
+	// batcher and the batches.
+	store BatcherStore
+
+	// wg is a waitgroup that is used to wait for all the goroutines to
+	// exit.
+	wg sync.WaitGroup
+}
+
+// NewBatcher creates a new Batcher instance.
+func NewBatcher(wallet lndclient.WalletKitClient,
+	chainNotifier lndclient.ChainNotifierClient,
+	signerClient lndclient.SignerClient, musig2ServerSigner MuSig2SignSweep,
+	verifySchnorrSig VerifySchnorrSig, chainparams *chaincfg.Params,
+	store BatcherStore) *Batcher {
+
+	return &Batcher{
+		batches:          make(map[int32]*batch),
+		SweepReqs:        make(chan SweepRequest),
+		errChan:          make(chan error),
+		completeChan:     make(chan int32),
+		wallet:           wallet,
+		chainNotifier:    chainNotifier,
+		signerClient:     signerClient,
+		musig2ServerSign: musig2ServerSigner,
+		VerifySchnorrSig: verifySchnorrSig,
+		chainParams:      chainparams,
+		store:            store,
+	}
+}
+
+// Run starts the batcher and processes incoming sweep requests.
+func (b *Batcher) Run(ctx context.Context) error {
+	runCtx, cancel := context.WithCancel(ctx)
+	defer func() {
+		cancel()
+
+		for _, batch := range b.batches {
+			batch.Wait()
+		}
+
+		b.wg.Wait()
+	}()
+
+	// First we fetch all the batches that are not in a confirmed state from
+	// the database. We will then resume the execution of these batches.
+	batches, err := b.FetchUnconfirmedBatches(runCtx)
+	if err != nil {
+		return err
+	}
+
+	for _, batch := range batches {
+		err := b.spinUpBatchFromDB(runCtx, batch)
+		if err != nil {
+			return err
+		}
+	}
+
+	for {
+		select {
+		case sweepReq := <-b.SweepReqs:
+			sweep, err := b.fetchSweep(runCtx, sweepReq)
+			if err != nil {
+				return err
+			}
+
+			err = b.handleSweep(runCtx, sweep, sweepReq.Notifier)
+			if err != nil {
+				return err
+			}
+
+		case id := <-b.completeChan:
+			delete(b.batches, id)
+
+		case err := <-b.errChan:
+			return err
+
+		case <-runCtx.Done():
+			return runCtx.Err()
+		}
+	}
+}
+
+// handleSweep handles a sweep request by either placing it in an existing
+// batch, or by spinning up a new batch for it.
+func (b *Batcher) handleSweep(ctx context.Context, sweep *sweep,
+	notifier SpendNotifier) error {
+
+	// When batches purge sweeps they will be returned here in order to be
+	// handled again. In that case the batch will not define a notifier as
+	// it does not care about the future outcome of the sweep. The swap has
+	// already registered a notifier with the batcher.
+	if notifier != (SpendNotifier{}) {
+		b.wg.Add(1)
+		go b.monitorSpendAndNotify(ctx, sweep, notifier)
```

Comment:
If the notifier is passed to the batch along with the sweep, then when it is spent we could just notify from there right? No real need for this extra monitor/sweep? That way the notifier would also be always set.

Reply:
Bumping this as I still think that there's no real need for this monitor given the batch's primary sweep is already monitored in the batch itself and we can notify the swap of both the publish and the confirm events.